automerge-c/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+        let handle = $handle.as_ref();
+        match handle {
+            Some(b) => b,
+            None => &AMobj(am::ObjId::Root),
+        }
+    }};
+}
+
+macro_rules! to_result {
+    ($val:expr) => {{
+        Box::into_raw(Box::new(($val).into()))
+    }};
+}
+
+/// \class AMobj
+/// \brief An object's unique identifier.
+#[derive(Clone)]
+pub struct AMobj(am::ObjId);
+
+impl AsRef<am::ObjId> for AMobj {
+    fn as_ref(&self) -> &am::ObjId {
+        &self.0
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Allocates a new `AMdoc` struct and initializes it with defaults.
+///
+/// \return A pointer to an `AMdoc` struct.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMdestroy()`.
+#[no_mangle]
+pub extern "C" fn AMcreate() -> *mut AMdoc {
+    AMdoc::create(am::AutoCommit::new()).into()
+}
+
+/// \memberof AMdoc
+/// \brief Deallocates the storage for an `AMdoc` struct previously
+///        allocated by `AMcreate()` or `AMdup()`.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \pre \p doc must be a valid address.
+#[no_mangle]
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+pub unsafe extern "C" fn AMdestroy(doc: *mut AMdoc) {
+    if !doc.is_null() {
+        let doc: AMdoc = *Box::from_raw(doc);
+        drop(doc)
+    }
+}
+
+/// \memberof AMdoc
+/// \brief Allocates storage for an `AMdoc` struct and initializes it by
+///        duplicating the `AMdoc` struct pointed to by \p doc.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \return A pointer to an `AMdoc` struct.
+/// \pre \p doc must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMdestroy()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+#[no_mangle]
+pub unsafe extern "C" fn AMdup(doc: *mut AMdoc) -> *mut AMdoc {
+    let doc = *Box::from_raw(doc);
+    let copy = doc.clone();
+    std::mem::forget(doc);
+    copy.into()
+}
+
+/// \memberof AMdoc
+/// \brief Sets a configuration property of an `AMdoc` struct.
+///
+/// \param[in] doc A pointer to an `AMdoc` struct.
+/// \param[in] key A configuration property's key string.
+/// \param[in] value A configuration property's string value or `NULL`.
+/// \return A pointer to an `AMresult` struct containing no value.
+/// \pre \p doc must be a valid address.
+/// \pre \p key must be a valid address.
+/// \warning To avoid a memory leak, the returned pointer must be deallocated
+///          with `AMclear()`.
+///
+/// # Safety
+/// doc must be a pointer to a valid AMdoc
+/// key and value must be valid c strings
+#[no_mangle]
+pub unsafe extern "C" fn AMconfig(
+    doc: *mut AMdoc,
+    key: *const c_char,
+    value: *const c_char,
+) -> *mut AMresult {
+    let doc = to_doc!(doc);
+    let key = to_str!(key);
+    match key.as_str() {
+        "actor" => {
+            let actor = to_str!(value);
+            if let Ok(actor) = actor.try_into() {
```

Comment:
> ActorId is just bytes under the covers and this conversion to a string feels like it could be really confusing because we replace any non unicode characters with U+FFFD.

@alexjg, that's interesting. Do you think the C function exposing the `types::ActorId::random()` function should only return `ActorId` values whose bytes correspond with Unicode characters?

Reply:
An actor id is a byte array of arbitrary size.  In the C api we should be treating these as `char *` with a length or a hex string.  Or both.
Thinking about it the `char*` makes the most sense since this is such a low level api - but I've always liked the convinces of hex strings for writing examples and test code.  Maybe we should support both.

Reply:
I think exposing both is a good idea, with a preference for the `char *` with length first on the basis that it's more flexible.

Reply:
> I think exposing both is a good idea, with a preference for the char * with length first on the basis that it's more flexible.

@alexjg, both options are implemented in PR #310.